Improve async session updates

joshblakeley opened the pull request:
Fixes #1758 

Current solution for session updates when async is enabled is just to spawn a new goroutine per redis operation. 

This change allows setting a configurable pool of workers to handle the session updates - with the slower parts of the session update such as json marshalling moved into the pool. Default value if unset is currently 15 and worker pool will gracefully exit on gateway shutdown.

Also changed the dummySessionManager in policy_test.go pass by ref to stop go vet failing because of passed mutex.

joshblakeley commented:
Currently panics on gateway shutdown so have that to fix still.

buger commented:
Can you put some initial benchmarks to understand the difference it makes?

joshblakeley commented:
Yes, will do and update top comment.

buger commented:
Also if it will show good performance, and we decide to turn async session write by default, it will make sense remove old setting, and introduce new one: `disable_async_session_writes`

joshblakeley commented:
```benchmark                                                   old ns/op     new ns/op     delta
BenchmarkURLReplacer-4                                      1558          1589          +1.99%
BenchmarkTagHeaders-4                                       817           819           +0.24%
BenchmarkDefaultVersion-4                                   1064771       1093007       +2.65%
BenchmarkGetVersionFromRequest/Header_location-4            482137        459508        -4.69%
BenchmarkGetVersionFromRequest/URL_param_location-4         458418        456370        -0.45%
BenchmarkGetVersionFromRequest/URL_location-4               459188        463693        +0.98%
BenchmarkApiReload-4                                        3948803       4107164       +4.01%
BenchmarkValueExtractor/HeaderSource-4                      620281        656742        +5.88%
BenchmarkValueExtractor/FormSource-4                        671565        660416        -1.66%
BenchmarkRegexExtractor/HeaderSource-4                      634194        663528        +4.63%
BenchmarkRegexExtractor/BodySource-4                        553221        604820        +9.33%
BenchmarkRegexExtractor/FormSource-4                        625323        689967        +10.34%
BenchmarkXPathExtractor/HeaderSource-4                      720379        706684        -1.90%
BenchmarkXPathExtractor/BodySource-4                        720888        694420        -3.67%
BenchmarkXPathExtractor/FormSource-4                        578117        695444        +20.29%
BenchmarkInitGenericEventHandlers-4                         703           703           +0.00%
BenchmarkMultiSession_BA_Standard_OK-4                      1730595       1927234       +11.36%
BenchmarkBearerTokenAuthKeySession-4                        1638861       1964912       +19.89%
BenchmarkMultiAuthBackwardsCompatibleSession-4              1794075       1918844       +6.95%
BenchmarkBasicAuth-4                                        134236255     133807972     -0.32%
BenchmarkContextVarsMiddleware-4                            1244750       1239212       -0.44%
BenchmarkMiddlewareContextVars_ProcessRequest_cookies-4     236467        240262        +1.60%
BenchmarkHMACAuthSessionPass-4                              1274908       1267094       -0.61%
BenchmarkHMACAuthSessionPassWithHeaderField-4               1209695       1199873       -0.81%
BenchmarkIPBlacklistMiddleware-4                            25786         25999         +0.83%
BenchmarkIPMiddlewarePass-4                                 16842         16903         +0.36%
BenchmarkJWTSessionHMAC-4                                   360199        351308        -2.47%
BenchmarkJWTSessionRSA-4                                    460825        452648        -1.77%
BenchmarkJWTSessionRSABearer-4                              457821        455720        -0.46%
BenchmarkJWTSessionRSAWithRawSourceOnWithClientID-4         494911        476905        -3.64%
BenchmarkJWTSessionRSAWithRawSource-4                       487164        463405        -4.88%
BenchmarkJWTSessionRSAWithJWK-4                             492495        463193        -5.95%
BenchmarkJWTSessionRSAWithEncodedJWK-4                      456223        461279        +1.11%
BenchmarkProcessRequestLiveQuotaLimit-4                     1658338       1868746       +12.69%
BenchmarkProcessRequestOffThreadQuotaLimit-4                1361758       1245086       -8.57%
BenchmarkProcessRequestLiveRedisRollingLimiter-4            1480984       1368393       -7.60%
BenchmarkProcessRequestOffThreadRedisRollingLimiter-4       1301096       1296460       -0.36%
BenchmarkStripAuth_stripFromHeaders-4                       7660          7585          -0.98%
BenchmarkStripAuth_stripFromParams-4                        35010         34868         -0.41%
BenchmarkTransformNonAscii-4                                14345         14529         +1.28%
BenchmarkTransformJSONMarshal-4                             15210         15456         +1.62%
BenchmarkRewriter-4                                         31763         32094         +1.04%
BenchmarkValidateJSONSchema-4                               3519864       3466755       -1.51%
BenchmarkVersioning-4                                       4547555       4435786       -2.46%
BenchmarkVirtualEndpoint-4                                  1147857       1074119       -6.42%
BenchmarkApplyPolicies-4                                    22216         22664         +2.02%
BenchmarkResponseHeaderInjection-4                          3969715       3818353       -3.81%
BenchmarkRequestIPHops-4                                    399           394           -1.25%
BenchmarkWrappedServeHTTP-4                                 127415        125208        -1.73%
BenchmarkCopyRequestResponse-4                              14770         14182         -3.98%
BenchmarkRegExpCompile-4                                    115           119           +3.48%
BenchmarkRegExpCompilePOSIX-4                               115           116           +0.87%
BenchmarkRegExpMustCompile-4                                116           118           +1.72%
BenchmarkRegExpMustCompilePOSIX-4                           116           117           +0.86%
BenchmarkRegExpMatchString-4                                201           201           +0.00%
BenchmarkRegExpMatch-4                                      206           207           +0.49%
BenchmarkRegExpString-4                                     0.86          0.85          -1.16%
BenchmarkRegexpReplaceAllString-4                           83.6          83.7          +0.12%
BenchmarkRegexpReplaceAllLiteralString-4                    83.9          84.7          +0.95%
BenchmarkRegexpReplaceAllStringFunc-4                       187           186           -0.53%
BenchmarkRegexpFindAllString-4                              110           111           +0.91%
BenchmarkRegexpFindAllStringSubmatch-4                      110           111           +0.91%
BenchmarkRealIP_RemoteAddr-4                                136           137           +0.74%
BenchmarkRealIP_ForwardedFor-4                              127           127           +0.00%
BenchmarkRealIP_RealIP-4                                    96.3          95.1          -1.25%
BenchmarkRealIP_Context-4                                   12.9          11.8          -8.53%
BenchmarkHash-4                                             3543          3547          +0.11%

benchmark                                                   old allocs     new allocs     delta
BenchmarkURLReplacer-4                                      12             12             +0.00%
BenchmarkTagHeaders-4                                       15             15             +0.00%
BenchmarkDefaultVersion-4                                   1410           1314           -6.81%
BenchmarkGetVersionFromRequest/Header_location-4            544            544            +0.00%
BenchmarkGetVersionFromRequest/URL_param_location-4         528            528            +0.00%
BenchmarkGetVersionFromRequest/URL_location-4               509            509            +0.00%
BenchmarkApiReload-4                                        23903          23903          +0.00%
BenchmarkValueExtractor/HeaderSource-4                      75             75             +0.00%
BenchmarkValueExtractor/FormSource-4                        102            102            +0.00%
BenchmarkRegexExtractor/HeaderSource-4                      77             77             +0.00%
BenchmarkRegexExtractor/BodySource-4                        86             87             +1.16%
BenchmarkRegexExtractor/FormSource-4                        104            104            +0.00%
BenchmarkXPathExtractor/HeaderSource-4                      133            133            +0.00%
BenchmarkXPathExtractor/BodySource-4                        141            141            +0.00%
BenchmarkXPathExtractor/FormSource-4                        165            165            +0.00%
BenchmarkInitGenericEventHandlers-4                         8              8              +0.00%
BenchmarkMultiSession_BA_Standard_OK-4                      468            468            +0.00%
BenchmarkBearerTokenAuthKeySession-4                        421            422            +0.24%
BenchmarkMultiAuthBackwardsCompatibleSession-4              428            428            +0.00%
BenchmarkBasicAuth-4                                        19901          19895          -0.03%
BenchmarkContextVarsMiddleware-4                            1710           1709           -0.06%
BenchmarkMiddlewareContextVars_ProcessRequest_cookies-4     28             28             +0.00%
BenchmarkHMACAuthSessionPass-4                              255            255            +0.00%
BenchmarkHMACAuthSessionPassWithHeaderField-4               274            273            -0.36%
BenchmarkIPBlacklistMiddleware-4                            152            152            +0.00%
BenchmarkIPMiddlewarePass-4                                 95             95             +0.00%
BenchmarkJWTSessionHMAC-4                                   512            512            +0.00%
BenchmarkJWTSessionRSA-4                                    614            614            +0.00%
BenchmarkJWTSessionRSABearer-4                              616            616            +0.00%
BenchmarkJWTSessionRSAWithRawSourceOnWithClientID-4         636            636            +0.00%
BenchmarkJWTSessionRSAWithRawSource-4                       626            626            +0.00%
BenchmarkJWTSessionRSAWithJWK-4                             629            629            +0.00%
BenchmarkJWTSessionRSAWithEncodedJWK-4                      634            634            +0.00%
BenchmarkProcessRequestLiveQuotaLimit-4                     589            590            +0.17%
BenchmarkProcessRequestOffThreadQuotaLimit-4                592            591            -0.17%
BenchmarkProcessRequestLiveRedisRollingLimiter-4            1474           1521           +3.19%
BenchmarkProcessRequestOffThreadRedisRollingLimiter-4       1543           1553           +0.65%
BenchmarkStripAuth_stripFromHeaders-4                       66             66             +0.00%
BenchmarkStripAuth_stripFromParams-4                        220            220            +0.00%
BenchmarkTransformNonAscii-4                                98             98             +0.00%
BenchmarkTransformJSONMarshal-4                             101            101            +0.00%
BenchmarkRewriter-4                                         238            238            +0.00%
BenchmarkValidateJSONSchema-4                               2672           2430           -9.06%
BenchmarkVersioning-4                                       2626           2625           -0.04%
BenchmarkVirtualEndpoint-4                                  464            464            +0.00%
BenchmarkApplyPolicies-4                                    102            102            +0.00%
BenchmarkResponseHeaderInjection-4                          2490           2248           -9.72%
BenchmarkRequestIPHops-4                                    4              4              +0.00%
BenchmarkWrappedServeHTTP-4                                 144            144            +0.00%
BenchmarkCopyRequestResponse-4                              76             76             +0.00%
BenchmarkRegExpCompile-4                                    2              2              +0.00%
BenchmarkRegExpCompilePOSIX-4                               2              2              +0.00%
BenchmarkRegExpMustCompile-4                                2              2              +0.00%
BenchmarkRegExpMustCompilePOSIX-4                           2              2              +0.00%
BenchmarkRegExpMatchString-4                                3              3              +0.00%
BenchmarkRegExpMatch-4                                      3              3              +0.00%
BenchmarkRegExpString-4                                     0              0              +0.00%
BenchmarkRegexpReplaceAllString-4                           1              1              +0.00%
BenchmarkRegexpReplaceAllLiteralString-4                    1              1              +0.00%
BenchmarkRegexpReplaceAllStringFunc-4                       2              2              +0.00%
BenchmarkRegexpFindAllString-4                              2              2              +0.00%
BenchmarkRegexpFindAllStringSubmatch-4                      2              2              +0.00%
BenchmarkRealIP_RemoteAddr-4                                1              1              +0.00%
BenchmarkRealIP_ForwardedFor-4                              1              1              +0.00%
BenchmarkRealIP_RealIP-4                                    1              1              +0.00%
BenchmarkRealIP_Context-4                                   0              0              +0.00%
BenchmarkHash-4                                             10             10             +0.00%

benchmark                                                   old bytes     new bytes     delta
BenchmarkURLReplacer-4                                      1088          1088          +0.00%
BenchmarkTagHeaders-4                                       384           384           +0.00%
BenchmarkDefaultVersion-4                                   199657        193743        -2.96%
BenchmarkGetVersionFromRequest/Header_location-4            85429         85410         -0.02%
BenchmarkGetVersionFromRequest/URL_param_location-4         85065         85042         -0.03%
BenchmarkGetVersionFromRequest/URL_location-4               82377         82352         -0.03%
BenchmarkApiReload-4                                        1788971       1788977       +0.00%
BenchmarkValueExtractor/HeaderSource-4                      5659          5666          +0.12%
BenchmarkValueExtractor/FormSource-4                        11763         11762         -0.01%
BenchmarkRegexExtractor/HeaderSource-4                      5702          5698          -0.07%
BenchmarkRegexExtractor/BodySource-4                        10139         10148         +0.09%
BenchmarkRegexExtractor/FormSource-4                        11794         11799         +0.04%
BenchmarkXPathExtractor/HeaderSource-4                      10472         10470         -0.02%
BenchmarkXPathExtractor/BodySource-4                        14891         14891         +0.00%
BenchmarkXPathExtractor/FormSource-4                        17072         17081         +0.05%
BenchmarkInitGenericEventHandlers-4                         528           528           +0.00%
BenchmarkMultiSession_BA_Standard_OK-4                      79027         79042         +0.02%
BenchmarkBearerTokenAuthKeySession-4                        74552         74619         +0.09%
BenchmarkMultiAuthBackwardsCompatibleSession-4              76278         76264         -0.02%
BenchmarkBasicAuth-4                                        1040143       1039650       -0.05%
BenchmarkContextVarsMiddleware-4                            281031        280881        -0.05%
BenchmarkMiddlewareContextVars_ProcessRequest_cookies-4     2401          2401          +0.00%
BenchmarkHMACAuthSessionPass-4                              49571         49460         -0.22%
BenchmarkHMACAuthSessionPassWithHeaderField-4               49203         49106         -0.20%
BenchmarkIPBlacklistMiddleware-4                            32312         32309         -0.01%
BenchmarkIPMiddlewarePass-4                                 26709         26708         -0.00%
BenchmarkJWTSessionHMAC-4                                   80361         80348         -0.02%
BenchmarkJWTSessionRSA-4                                    102385        102386        +0.00%
BenchmarkJWTSessionRSABearer-4                              103586        103591        +0.00%
BenchmarkJWTSessionRSAWithRawSourceOnWithClientID-4         104424        104420        -0.00%
BenchmarkJWTSessionRSAWithRawSource-4                       101847        101827        -0.02%
BenchmarkJWTSessionRSAWithJWK-4                             100824        100814        -0.01%
BenchmarkJWTSessionRSAWithEncodedJWK-4                      101044        101039        -0.00%
BenchmarkProcessRequestLiveQuotaLimit-4                     84135         84171         +0.04%
BenchmarkProcessRequestOffThreadQuotaLimit-4                84152         84106         -0.05%
BenchmarkProcessRequestLiveRedisRollingLimiter-4            115604        117429        +1.58%
BenchmarkProcessRequestOffThreadRedisRollingLimiter-4       118319        118721        +0.34%
BenchmarkStripAuth_stripFromHeaders-4                       2800          2800          +0.00%
BenchmarkStripAuth_stripFromParams-4                        14162         14162         +0.00%
BenchmarkTransformNonAscii-4                                11059         11059         +0.00%
BenchmarkTransformJSONMarshal-4                             11507         11507         +0.00%
BenchmarkRewriter-4                                         45597         45597         +0.00%
BenchmarkValidateJSONSchema-4                               280151        269032        -3.97%
BenchmarkVersioning-4                                       268537        268541        +0.00%
BenchmarkVirtualEndpoint-4                                  71799         71805         +0.01%
BenchmarkApplyPolicies-4                                    12582         12582         +0.00%
BenchmarkResponseHeaderInjection-4                          385162        373856        -2.94%
BenchmarkRequestIPHops-4                                    432           432           +0.00%
BenchmarkWrappedServeHTTP-4                                 48389         48389         +0.00%
BenchmarkCopyRequestResponse-4                              35585         35585         +0.00%
BenchmarkRegExpCompile-4                                    176           176           +0.00%
BenchmarkRegExpCompilePOSIX-4                               176           176           +0.00%
BenchmarkRegExpMustCompile-4                                176           176           +0.00%
BenchmarkRegExpMustCompilePOSIX-4                           176           176           +0.00%
BenchmarkRegExpMatchString-4                                192           192           +0.00%
BenchmarkRegExpMatch-4                                      192           192           +0.00%
BenchmarkRegExpString-4                                     0             0             +0.00%
BenchmarkRegexpReplaceAllString-4                           16            16            +0.00%
BenchmarkRegexpReplaceAllLiteralString-4                    16            16            +0.00%
BenchmarkRegexpReplaceAllStringFunc-4                       48            48            +0.00%
BenchmarkRegexpFindAllString-4                              34            34            +0.00%
BenchmarkRegexpFindAllStringSubmatch-4                      34            34            +0.00%
BenchmarkRealIP_RemoteAddr-4                                16            16            +0.00%
BenchmarkRealIP_ForwardedFor-4                              16            16            +0.00%
BenchmarkRealIP_RealIP-4                                    16            16            +0.00%
BenchmarkRealIP_Context-4                                   0             0             +0.00%
BenchmarkHash-4                                             1360          1360          +0.00%
```

dencoded commented on auth_manager.go:
```diff
@@ -89,6 +103,56 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+
+		log.WithField("Auth Manager", b.poolSize).Debug("Session update async pool size")
+
+		b.updateChan = make(chan *SessionUpdate)
```
I think we need to have that channel buffered with buffer size specified in config

joshblakeley replied:
What is a sensible default for this?

joshblakeley replied:
Using 1000 as default

dencoded commented on auth_manager.go:
```diff
@@ -89,6 +103,56 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+		if u.isHashed {
+			u.keyVal = b.store.GetKeyPrefix() + u.keyVal
+			b.store.SetRawKey(u.keyVal, string(v), u.ttl)
+
```
is `continue` missed here?

dencoded commented on auth_manager.go:
```diff
@@ -89,6 +103,56 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+		v, _ := json.Marshal(u.session)
+
+		if u.isHashed {
+			u.keyVal = b.store.GetKeyPrefix() + u.keyVal
```
I'd read key prefix at Init() time and store it in one of the fields, then use here to eliminate one function call

dencoded commented on auth_manager.go:
```diff
@@ -89,6 +103,56 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+
+		if u.isHashed {
+			u.keyVal = b.store.GetKeyPrefix() + u.keyVal
+			b.store.SetRawKey(u.keyVal, string(v), u.ttl)
```
we need to catch and log error here

dencoded commented on auth_manager.go:
```diff
@@ -89,6 +103,56 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+
+		}
+
+		b.store.SetKey(u.keyVal, string(v), u.ttl)
```
we need to catch and log error here as well

dencoded commented on auth_manager.go:
```diff
@@ -89,6 +103,56 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+		if !ok {
+			return
+		}
+		v, _ := json.Marshal(u.session)
```
we need to catch and log marshalling error here, continue if it was errored

dencoded commented on auth_manager.go:
```diff
@@ -89,6 +103,56 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+func (b *DefaultSessionManager) updateWorker() {
+	defer b.poolWG.Done()
+
+	for {
```
here probably here would be better to use `range` cycle instead of `select` as we don't have to read from several channels, i.e.:
```
for u := range b.updateChan {
    ....
}
``` 
cycle will exit automatically once channel is closed

dencoded commented on auth_manager.go:
```diff
@@ -89,6 +103,56 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+}
+
+func (b *DefaultSessionManager) Stop() {
+	//flag to stop adding data to chan
```
spaces in in-line comments after `//`

dencoded commented on auth_manager.go:
```diff
@@ -120,25 +184,32 @@ func (b *DefaultSessionManager) UpdateSession(keyName string, session *user.Sess
-		if hashed {
-			go b.store.SetRawKey(keyName, string(v), resetTTLTo)
-			return nil
+		sessionMeta := &SessionUpdate{
```
I am bad at naming but I feel like it could be just `sessionUpdate` (not meta)

dencoded commented on auth_manager.go:
```diff
@@ -120,25 +184,32 @@ func (b *DefaultSessionManager) UpdateSession(keyName string, session *user.Sess
 		return nil
 	}
 
+	v, _ := json.Marshal(session)
```
we need catch, log error here and return if it was error

dencoded commented on auth_manager.go:
```diff
@@ -120,25 +184,32 @@ func (b *DefaultSessionManager) UpdateSession(keyName string, session *user.Sess
-			go b.store.SetRawKey(keyName, string(v), resetTTLTo)
-			return nil
+		sessionMeta := &SessionUpdate{
+			isHashed: hashed,
```
we could eliminate this field if we pass just keyName (we know here if prefix needs to be added)

dencoded commented on main.go:
```diff
@@ -1052,11 +1053,27 @@ func main() {
 	mainLog.Info("Stop signal received.")
 
+	//if using async session writes stop workers
+	if config.Global().UseAsyncSessionWrite {
```
there will be also two pools of workers (`spec.SessionManager` and `spec.OrgSessionManager`) running per each API spec started here https://github.com/TykTechnologies/tyk/blob/v2.6.2/api_loader.go#L188 so we need to stop them as well before exit. Maybe just go through all loaded specs and stop pools

dencoded commented on main.go:
```diff
@@ -1052,11 +1053,27 @@ func main() {
+	//if using async session writes stop workers
+	if config.Global().UseAsyncSessionWrite {
+		DefaultOrgStore.Stop()
+		DefaultQuotaStore.Stop()
```
we do  only `DefaultQuotaStore.RemoveSession` so why does it need pool?

I am not even sure if `DefaultQuotaStore` is needed any more, I wasn't able to find its usage, might worth to double check if it is populated in Redis by some 3d party

joshblakeley commented:
It's ready for further review i just get a strange unit test failure im trying to work out at the moment.

joshblakeley commented:
@buger I fixed CI but in a very hacky way - basically the test that used the sentinel rate limiter would fail its cases unless i added some extra lines.

That test has nothing to do with my code but there must be something on the memory level affecting it i think.

buger commented:
@josephearl btw when you fix such things better push as separate commit next time, so it will be easy to track PR evolution and such PR fixes. We can always squash it in the end.

buger commented:
@dencoded can you pls re-review?

joshblakeley commented:
Ok will do so in future!

joshblakeley commented:
@dencoded bump for review

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+			b.bufferSize = 1000
+		}
+
+		log.WithField("Auth Manager", b.poolSize).Debug("Session update async pool size")
```
`WithField` needs to be used with some meaningful field name

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+		// check pool size in config and set to 50 if unset
+		b.poolSize = config.Global().SessionUpdatePoolSize
+		if b.poolSize <= 0 {
+			b.poolSize = 50
```
how about using constant for `50`

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+		//check size for channel buffer and set to 1000 if unset
+		b.bufferSize = config.Global().SessionUpdateBufferSize
+		if b.bufferSize <= 0 {
+			b.bufferSize = 1000
```
the same question about using constant for `1000`

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+func (b *DefaultSessionManager) updateWorker() {
+	defer b.poolWG.Done()
+
+	for range b.updateChan {
```
`for u := range b.updateChan`

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+
+	for range b.updateChan {
+		// grab update object from channel
+		u := <-b.updateChan
```
we don't need to do this when for-range is used to read from channel

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+
+		v, err := json.Marshal(u.session)
+		if err != nil {
+			log.Error("Error marshalling session for async session update")
```
`log.WithError(err).Error("Error marshalling session for async session update")` would be more informative

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+		}
+
+		if u.isHashed {
+			u.keyVal = b.keyPrefix + u.keyVal
```
why do we change passed instance of `SessionUpdate` here?

joshblakeley replied:
This is just the same as what we did in existing implementation. When we call the setRawKey method to update Redis it expects the key to be prefixed probably because this is what we do on hashed key creation for that keyID.

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+			u.keyVal = b.keyPrefix + u.keyVal
+			err := b.store.SetRawKey(u.keyVal, string(v), u.ttl)
+			if err != nil {
+				log.Errorf("Error updating hashed key: %v", err)
```
logrus offers api for outputting errors as a special field `log.WithError(err).Error("Error updating hashed key")`

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+
+		err = b.store.SetKey(u.keyVal, string(v), u.ttl)
+		if err != nil {
+			log.Errorf("Error updating non-hashed key: %v", err)
```
the same note about `log.WithError(err)....`

dencoded commented on auth_manager.go:
```diff
@@ -86,6 +103,73 @@ func (b *DefaultSessionManager) Init(store storage.Handler) {
+}
+
+func (b *DefaultSessionManager) Stop() {
+	if atomic.LoadUint32(&b.shouldStop) == 0 {
```
being fun of earlier return I would do
```
if atomic.LoadUint32(&b.shouldStop) == 1 {
    return
}
... // stop logic
```

dencoded commented on auth_manager.go:
```diff
@@ -111,25 +195,37 @@ func (b *DefaultSessionManager) ResetQuota(keyName string, session *user.Session
+		return err
+	}
+
+	if hashed {
```
two `if hashed` in a row looks a little confusing

joshblakeley commented:
@dencoded bump

dencoded commented on auth_manager.go:
```diff
@@ -40,6 +40,9 @@ type SessionHandler interface {
 	Stop()
 }
 
+const sessionPoolDefault = 50
```
maybe word `size` is missing in constants names
